pfsspy/utils.py

```diff
@@ -0,0 +1,142 @@
+import sunpy.map,sunpy.io
+import numpy as np
+import os, warnings,copy
+from scipy import interpolate
+from astropy import units as u
+
+def roll_to_CRLN(synmap) :
+    """ Roll sunpy.map.Map synoptic map so LH edge is 0 Carr Lon.
```

Comment:
This doesn't currently account for the case where 0 degrees of longitude lies in the middle of a pixel, and therefore some sort of re-binning would have to be done to make 0 degrees lie on the edge of a pixel as required. A much easier (and potentially less useful, but fine for GONG) alternative would be `pixel_roll(synmap, npix)`, that would roll the data by `npix` pixels, and just update the `CRVAL` appropriately, without having to deal with fractional pixels.

This should also check that the CTYPE of the map is appropriate (CRLN I think?) and error if not.

Reply:
So after staring at the FITS standards for a while, I think I was missing some subtlety on the meaning of CRPIX, but think I understand it now, could you check me on this:

* CRPIXn labels pixels and ranges from 1 to NAXISn. 
* Integer values of CRPIX mark the **center** of each pixel, i.e. they label a distinct column or row of pixels
* Therefore the left edge of the first pixel would actually be described as CRPIX = 0.5, and the right hand hedge of the last pixel (NAXISn) is NAXISn + 0.5
* For this application we therefore want to align Carrington 0 longitude to CRPIX = 0.5
* At least for GONG maps, the reference pixel in longitude is 180.5 which is a **pixel edge**, and references an integer number of pixels, therefore an integer roll can correctly take Carr 0 to CRPIX=0.5

Happy to take a shot at doing a roll to this condition in general but will fall back on your pixel_roll idea if I run into problems.

Reply:
> So after staring at the FITS standards for a while

Join the club, I've just spent the afternoon clarifying all of this pixel indexing in sunpy... I agree with all of the above points!

Reply:
I think I've accomplished the general case with https://github.com/dstansby/pfsspy/pull/173/commits/6a6022ec1041313f7b42fe3d0739a2fa04e1579b , let me know if you see any more gotchas!

The build is failing with sunpy master branch because of the test :

`test_pfss.test_header_generation`

which runs :  `pfsspy.carr_cea_wcs_header`

which runs `sunpy.map.make_fitswcs_header`

 `pfsspy.carr_cea_wcs_header` currently supplies the keyword `reference_pixel` explicitly referenced to the fits standard, whereas `sunpy.map.make_fitswcs_header` has now been updated to expect the reference pixel to be 0 referenced, and generates the "CRPIX1" keywords accordingly. 

I guess the future-facing thing to do is to change `pfsspy.carr_cea_wcs_header` to supply 0 referenced `reference_pixel` but I guess this shouldn't be done until the next official sunpy release?

The change would be in `pfsspy.__init__.py`
```python
 ln 132      reference_pixel=[(shape[1] / 2) + 0.5, (shape[0] / 2) + 0.5] * u.pix,  #(current)
 ln 132      reference_pixel=[(shape[1] / 2) - 0.5, (shape[0] / 2) - 0.5] * u.pix, # (update)
```

(sorry not sure how to do these snippets...)

Reply:
Thanks, will take a look! The test failure is my fault as it's a bug I fixed in SunPy 😆. https://github.com/dstansby/pfsspy/pull/193 should fix the test.